src/Transform/LowerToLLVM.cpp

```diff
@@ -60,6 +62,164 @@ static size_t getRankFromMemRefType(LLVM::LLVMType memRefTy) {
+    auto type = op->getResult(0).getType();
+    auto memRefTy = type.cast<mlir::MemRefType>();
+    auto llvmMemRefType =
+        typeConverter.convertType(type).cast<LLVM::LLVMType>();
+
+    // The element type of the array.
+    auto constantElementType =
+        typeConverter.convertType(memRefTy.getElementType());
+    auto globalType = constantElementType;
+    for (int i=shape.size() - 1; i >= 0; i--)
+      globalType = LLVM::LLVMType::getArrayTy(
+          globalType.cast<LLVM::LLVMType>(), ArrayAttrIntVal(shape, i));
+    // The llvm type of the global (example: [2 x [8 x float]])
+    auto llvmGlobalType = globalType.cast<LLVM::LLVMType>();
+
+    {
+      OpBuilder::InsertionGuard insertGuard(rewriter);
+      rewriter.setInsertionPointToStart(module.getBody());
+
+      global = rewriter.create<LLVM::GlobalOp>(loc,
+          llvmGlobalType, /*isConstant=*/true,
+          LLVM::Linkage::Internal, name, krnlGlobalOp.value());
+    }
+
+    // Some frequently used types.
+    auto llvmI8PtrTy = LLVM::LLVMType::getInt8PtrTy(llvmDialect);
+    auto llvmI64Ty = LLVM::LLVMType::getInt64Ty(llvmDialect);
+
+    // Allocate the memory where the constants will be used from.
+    // This is a region of local memory and needs to be emitted as an alloca.
+    auto one = rewriter.create<LLVM::ConstantOp>(loc,
+        llvmI64Ty, rewriter.getI64IntegerAttr(1));
+    auto alloc = rewriter.create<LLVM::AllocaOp>(
+        loc, llvmGlobalType.getPointerTo(), one, /*alignment=*/0);
+
+    // Copy constant value into the local alloca:
+    //  - Bitcast alloc to i8*
+    Value int8PtrAlloc = rewriter.create<LLVM::BitcastOp>(
+        loc, llvmI8PtrTy, alloc);
+    //  - Bitcast global to i8*
+    Value globalValue = rewriter.create<LLVM::AddressOfOp>(loc, global);
+    Value i8PtrGlobal = rewriter.create<LLVM::BitcastOp>(
+        loc, llvmI8PtrTy, globalValue);
+    //  - Set size.
+    Value memRefElementSize = rewriter.create<LLVM::ConstantOp>(loc,
+        llvmI64Ty, rewriter.getI64IntegerAttr(
+            getMemRefEltSizeInBytes(memRefTy)));
+    Value numElementsValue = rewriter.create<LLVM::ConstantOp>(
+        loc, llvmI64Ty, rewriter.getI64IntegerAttr(numElements));
+    Value totalElementsSize = rewriter.create<LLVM::MulOp>(
+        loc, memRefElementSize, numElementsValue);
+    Value int64Size = rewriter.create<LLVM::SExtOp>(
+        loc, llvmI64Ty, totalElementsSize);
+    //  - Set volatile.
+    Value isVolatile = rewriter.create<LLVM::ConstantOp>(
+        loc, LLVM::LLVMType::getInt1Ty(llvmDialect),
+        rewriter.getIntegerAttr(rewriter.getIntegerType(1), 0));
+    //  - Copy constant data into the alloca.
+    auto memcpyRef = getOrInsertMemcpy(rewriter, module, llvmDialect);
+    rewriter.create<CallOp>(
+        loc, memcpyRef, LLVM::LLVMType::getVoidTy(llvmDialect),
+        ArrayRef<Value>({int8PtrAlloc, i8PtrGlobal, int64Size, isVolatile}));
+
+    // Create the llvm.mlir.undef corresponding to the MemRef.
```

Comment:
Lines 190--211 can be replaced by
```mlir
auto llvmConstantElementType = constantElementType.cast<LLVM::LLVMType>();                                                                                             
Value typedAlloc = rewriter.create<LLVM::BitcastOp>(                                                                                                                   
    loc, llvmConstantElementType.getPointerTo(), alloc);                                                                                                               
auto llvmMemRef = MemRefDescriptor::fromStaticShape(                                                                                                                   
    rewriter, loc, typeConverter, memRefTy, typedAlloc); 
```
which helps create a MemRefType.

Reply:
I updated this. Thanks for pointing this out.